### ISSUE TITLE
Uncaught TypeError: Cannot read property '0' of undefined | Occurs every time you scroll

### ISSUE THREAD

**Issue by mitzey234** [atom/atom#18415]
[Enter steps to reproduce:]

1. Open some file with code, the files that cause this for me are .js
2. Start collapsing a couple or ALL functions or objects so you see all 500 lines or so reduced to about the appearance of 50 lines.
3. Start messing with the code a bit, for me things start getting sketchy where some lines are longer than the IDE is interpreting. Meaning I click my mouse cursor at the end of a line, but the IDE will only let me move the keyboard cursor to somewhere in between the end of the line and the front and will only let me move to the next line and not the remainder of that line. This is fixed by editing the line a bit. (Shown here: https://gfycat.com/IllinformedFemaleBubblefish)
4. Eventually somehow this leads into scrolling up or down in the editor to spit out an endless stream of errors like this one. (Shown Here: https://gfycat.com/GlumMassiveChipmunk)

**Atom**: 1.32.2 x64
**Electron**: 2.0.9
**OS**: Mac OS X 10.11.6
**Thrown From**: Atom Core

### Stack Trace

Uncaught TypeError: Cannot read property '0' of undefined

```
At /Applications/Atom.app/Contents/Resources/app/static/<embedded>:14

TypeError: Cannot read property '0' of undefined
    at DisplayLayer.collapseHardTabs (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:14:87524)
    at DisplayLayer.translateScreenPosition (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:14:85217)
    at DisplayMarkerLayer.t.exports.DisplayMarkerLayer.translateToBufferMarkerLayerFindParams (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:14:1099894)
    at DisplayMarkerLayer.t.exports.DisplayMarkerLayer.findMarkers (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:14:1097459)
    at decorationCountsByLayer.forEach (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:11:514669)
    at Map.forEach (<anonymous>)
    at DecorationManager.decorationPropertiesByMarkerForScreenRowRange (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:11:514643)
    at TextEditorComponent.queryDecorationsToRender (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:11:81701)
    at TextEditorComponent.updateSyncBeforeMeasuringContent (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:11:69794)
    at TextEditorComponent.updateSync (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:11:67306)
    at TextEditorComponent.didMouseWheel (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:11:92055)
```

### Commands

```
  8x -2:16.1.0 core:move-up (input.hidden-input)
  2x -2:09.4.0 core:move-right (input.hidden-input)
  3x -2:08.8.0 core:move-left (input.hidden-input)
  2x -2:07.8.0 core:move-right (input.hidden-input)
     -2:06.5.0 core:move-left (input.hidden-input)
     -2:06.3.0 core:backspace (input.hidden-input)
  4x -2:04.7.0 core:move-right (input.hidden-input)
     -1:59.4.0 core:move-left (input.hidden-input)
     -1:59.1.0 core:backspace (input.hidden-input)
  2x -1:58.2.0 core:save (input.hidden-input)
     -1:34 core:backspace (input.hidden-input)
  7x -1:33.6.0 core:move-right (input.hidden-input)
  3x -1:32 core:backspace (input.hidden-input)
  2x -1:30.4.0 core:save (input.hidden-input)
  2x -1:20.1.0 core:backspace (input.hidden-input)
  2x -1:16.5.0 core:copy (input.hidden-input)
```

### Non-Core Packages

```
atom-clock 0.1.16 
npm-install 4.0.4 
platformio-ide-terminal 2.8.4 
```

**Comment by rsese** [comment]
Thanks for the report!  It sounds like this issue is reproducible for you - just to confirm, can you reproduce in [safe mode](https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode)?

And about your steps to reproduce, can you share a file where this reproduces for you so we can try test things out on our end?

**Comment by mitzey234** [comment]
I'm still trying to work out if this is something that can be reproduced on demand and whether its something to do with the file I'm working on. Just now I managed to freeze the editor while performing step 2.

**Comment by mitzey234** [comment]
Ok. So this is going to require some crazy explanation. If I had to make an educated guess on what's going on here, I think it has something to do with saved-states or cached data about the individual file I'm working on. To explain this further, just earlier today I was doing some massive file transfer tasks using node.js via the Atom editor. At one point I pushed something way too far and it completely halted my kernel, I don't know how I managed to do that and I'm not sure how to explain why my OS was still working but couldn't launch anything. Since Atom was also running the node program, it along with other Electron or chromium based programs locked up.

I was forced to kill ALL chromium based programs from Discord to Atom to google chrome itself. In the end my OS never recovered and sudo reboot didn't do anything. So I pulled the plug and jumped back in. But what I suspect is happening now is temporary or cached information about this one file I'm working on got saved and damaged during this massive crash and once it was loaded in Atom, it caused some confusion for the editor. The reason I suspect this is because if I open a brand new window of Atom and take the exact same code and try reproducing the problem, it doesn't happen anymore. But if I use the old window that was recovered from the crash, this issue opens itself up and does things I didn't think were possible. 

Bottom line is, some weird things happened, and it's probably specific to my case, cause this is likely NOT something you can reproduce. Also blame javascript weirdness. 

I'm not sure if this is helpful in any way for anyone but my plan is to just close that window, reopen my project folder, and move on.

**Comment by mitzey234** [comment]
Nevermind I'm wrong, I thought I could move on and then this happened in safe mode.
https://gfycat.com/CelebratedNiftyBalloonfish

**Comment by rsese** [comment]
Thanks for the details @mitzey234 :+1:

I have seen some odd cases where saved state for a particular project gets messed up - if you don't have any unsaved data in any of your projects, you could [clear saved state](https://flight-manual.atom.io/hacking-atom/sections/debugging/#clearing-saved-state) and see if that helps.

And thanks for sharing that file and GIF - it looks like you're showing that the bottom of the pane isn't rendering your file till you scroll down?  If so, I tried to reproduce based on your GIF but couldn't:

1. fold at line 17
2. fold at line 7
3. scroll down

Does clearing saved state help in that case?

**Comment by mitzey234** [comment]
Cleared the save states, Problem still remains.

**Comment by Ben3eeE** [comment]
@rsese I can reproduce the folding issue. I reported it in another issue a while ago with a different file. /cc: https://github.com/atom/atom/issues/16078#issuecomment-370791712

My investigation back then was that cursor position mattered.

I was able to reproduce this issue with the file shared here as well.

**Comment by rsese** [comment]
:zap:

> My investigation back then was that cursor position mattered.

Ahhh yeah - I can reproduce with the file shared here too (and with yours from https://github.com/atom/atom/issues/16078#issuecomment-370791712).   I bet when I was trying to reproduce before that I clicked somewhere so didn't leave the cursor at the bottom of the file after copy/pasting the content to test with.

Update the issue body ✏️

**Comment by nathansobo** [comment]
@rsese Just wanted to note that @Ben3eeE reached out to me about this one. I'd like to evaluate prioritizing it in our next meeting.

**Comment by rsese** [comment]
Added for discussion :+1:
